File: haskell/problem_179.hs
-- Find the number of integers 1 < n < 10^(7), for which n and n + 1 have the same
-- number of positive divisors. For example, 14 has the positive divisors 1, 2, 7,
-- 14 while 15 has 1, 3, 5, 15.
-- 
-- http://projecteuler.net/index.php?section=problems&id=179

module Main where

import Data.Array.Unboxed
import Sigma

import Control.Parallel
import Control.Parallel.Strategies

mx = 10^5 - 1
-- mx = 25

-- sigmas :: [Int]
sigmas = tail $ foldl1 (\a b -> zipWith (+) a b) 
              $ [cycle 1:(take n $ repeat 0) | n <- [2..mx]]

main :: IO ()
main = do mapM_ (print) sigmas
          -- print $ sum $ zipWith (\a b -> if snd a == snd b then 1 else 0) sigmas $ tail sigmas

File: ruby/problem_179.rb
#!/usr/bin/env ruby -wKU
# 
# Find the number of integers 1 < n < 10^(7), for which n and n + 1 have the
# same number of positive divisors. For example, 14 has the positive divisors
# 1, 2, 7, 14 while 15 has 1, 3, 5, 15.
# 
# http://projecteuler.net/index.php?section=problems&id=179

require 'pp'

max = 10 ** 7 - 1

sigmas = []
(2..(max/2)).each do |interval|
  (1..(max/interval)).each do |index|
    sigmas[index * interval] ||= 0
    sigmas[index * interval] += 1
  end
end

n = 0
sigmas.each_with_index do |s, i|
  if s == sigmas[i + 1] then
    n += 1
  end
end
p n